### PR TITLE
Add support to optionally enable PhysX 5.1 in PhysX gem

### PR DESCRIPTION
**Pull request by SergeyAMZN**
## What does this PR do?

Add support to optionally enable PhysX 5.1 in PhysX gem via `-DAZ_USE_PHYSX5=ON` passed into CMake project generation command. 

RFC: https://github.com/o3de/sig-simulation/issues/31

## How was this PR tested?

QA pass, unit tests.

### CODE REVIEW

File: Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.h
Comment (moraaar) [inline Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.h:14]: This include, which I guess defines `PX_PHYSICS_VERSION_MAJOR` is only included here in `PhysXSceneQueryHelpers.h`, but the macro is used in many different cpps. Is it gurantee that all the cpps using the macro reaches `PhysXSceneQueryHelpers.h`? otherwise for some cpps `#if (PX_PHYSICS_VERSION_MAJOR == 5)` might be false and others true.

Maybe `#include <PxPhysicsVersion.h>` could be put in some basic header inside PhysX Gem to guarantee that?

File: cmake/3rdParty/Platform/Android/BuiltInPackages_android.cmake
Comment (moraaar) [inline cmake/3rdParty/Platform/Android/BuiltInPackages_android.cmake:34]: Since Android uses the script `generate_android_project.py` on top of CMake, currently there is no way to pass `-DAZ_USE_PHYSX5=ON` option to enable PhysX 5 on Android. We have to add an input parameter to `generate_android_project.py` and modify the scripts to be able to use PhysX 5 on Android.

File: Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.h
Comment (SergeyAMZN) [reply Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.h:14]: Good idea to try a non-unity build, all working fine with no issues.

There's no single basic header that is included everywhere (e.g. a precompiled header) but any code actually using PhysX SDK will have `PxPhysicsAPI.h` included which does include `PxPhysicsVersion.h` internally.

**Review by lawsonamzn — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h:76]: question: As `MeshMultiple` has changed from `(1 << 5)` to `(1 << 6)`, is there any chance this could impact or break any serialized data? Do we need to think of any kind of upgrade/conversion path or is this relatively safe you think?

File: Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h
Comment (hultonha) [inline Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h:79]: update: Could you please create a deprecation issue to track this (there's a template here - https://github.com/o3de/o3de/issues/new?assignees=&labels=kind%2Fdeprecation&template=deprecation_template.md&title=Deprecation+Notice). Then add the GHI here and we can then actually delete this after the next release goes out.

File: Gems/PhysX/Code/Source/BaseColliderComponent.cpp
Comment (hultonha) [inline Gems/PhysX/Code/Source/BaseColliderComponent.cpp:72]: nit: I believe we generally like to include closing comments for defines (so this would have `#endif // PX_PHYSICS_VERSION_MAJOR == 5`)

File: Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp
Comment (hultonha) [inline Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp:161]: Nice improvement! 😄

File: Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp
Comment (hultonha) [inline Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp:317]: update: Same comment before about the deprecation notice (could be the same one for all related code to make things easier to track)

File: Gems/PhysX/Code/Source/Pipeline/MeshExporter.cpp
Comment (hultonha) [inline Gems/PhysX/Code/Source/Pipeline/MeshExporter.cpp:139]: question: Should this only shift by 1 or 2 depending on if it's defined or not so we don't jump a large number of versions, won't the version number now be 37 (5 + 32)?

File: cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake
Comment (hultonha) [inline cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake:50]: nit/minor: Feel free to ignore but CMake has the `option` command (instead of set) to make this sort of thing simpler (see https://cmake.org/cmake/help/latest/command/option.html)

**Review by hultonha — COMMENTED** [review]
Overall looks good but I have a couple of questions before adding an approval. Great progress!

File: Gems/PhysX/Code/Source/BaseColliderComponent.cpp
Comment (lumberyard-employee-dm) [reply Gems/PhysX/Code/Source/BaseColliderComponent.cpp:72]: This is not needed for simple cases where the `#endif` is close to the `#if`: https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#comments-after-preprocessor-statements

File: cmake/3rdParty/Platform/iOS/BuiltInPackages_ios.cmake
Comment (lumberyard-employee-dm) [inline cmake/3rdParty/Platform/iOS/BuiltInPackages_ios.cmake:39]: These ly_associate_package calls can be moved to the PhysX gem itself, so that if the PhysX gem isn't part of the project, it can skip downloading the libraries.

Realistically though, my suggested change will not make a difference since the PhysX gem is part of the engine via the `engine.json`, but if the gem was ever not associated with the engine, but instead a project, then it would become applicable

**Review by lumberyard-employee-dm — COMMENTED** [review]
How long are we planning to support PhysX 4 after PhysX 5 support added?

Because all the split `#if (PX_PHYSICS_VERSION_MAJOR == 5)` defines should go away when PhysX 4 support is dropped.

I think it would be good to perhaps add an issue for the SIG-Simulation group to track how long PhysX 4 support would last

File: cmake/3rdParty/Platform/Android/BuiltInPackages_android.cmake
Comment (lumberyard-employee-dm) [reply cmake/3rdParty/Platform/Android/BuiltInPackages_android.cmake:34]: There is actually a way to pass input parameters to CMake via `generate_android_project.py`
The [--extra-cmake-configure-args](https://github.com/o3de/o3de/blob/development/cmake/Tools/Platform/Android/generate_android_project.py#L242-L244) argument can be used to supply arguments to the CMake.

**Comment by hultonha** [comment]
> How long are we planning to support PhysX 4 after PhysX 5 support added?
> 
> Because all the split `#if (PX_PHYSICS_VERSION_MAJOR == 5)` defines should go away when PhysX 4 support is dropped.
> 
> I think it would be good to perhaps add an issue for the SIG-Simulation group to track how long PhysX 4 support would last

+1 a tracking issue for PhysX 4 deprecation would be good... I don't think we have a firm ETA at this time but we want to ensure performance is still solid across mobile platforms that we're less sure of with PhysX 5. If all concerns are addressed there, then it's not unreasonable to assume PhysX 4 would be removed in development after the next release, but we're not committing to that at this stage, so it may be around longer.

File: Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h
Comment (SergeyAMZN) [reply Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h:76]: These are not serialized anywhere. 
Obviously technically external code could keep the hit flags saved somewhere in the database or file, but that'd be outside of our reflection system.

File: cmake/3rdParty/Platform/iOS/BuiltInPackages_ios.cmake
Comment (SergeyAMZN) [reply cmake/3rdParty/Platform/iOS/BuiltInPackages_ios.cmake:39]: That's a good idea, when it comes to supporting multiple default physics backends we won't have to have PhysX enabled. 
At the moment we can have more than one physics engine simulating, but the actual API backend is always PhysX.

File: cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake
Comment (SergeyAMZN) [reply cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake:50]: nice, looks like `option` would work too! 
I couldn't find any single use of `option()` in our cmake files though, so probably best to keep `set` for consistency?

**Comment by lumberyard-employee-dm** [comment]
> > How long are we planning to support PhysX 4 after PhysX 5 support added?
> > Because all the split `#if (PX_PHYSICS_VERSION_MAJOR == 5)` defines should go away when PhysX 4 support is dropped.
> > I think it would be good to perhaps add an issue for the SIG-Simulation group to track how long PhysX 4 support would last
> 
> +1 a tracking issue for PhysX 4 deprecation would be good... I don't think we have a firm ETA at this time but we want to ensure performance is still solid across mobile platforms that we're less sure of with PhysX 5. If all concerns are addressed there, then it's not unreasonable to assume PhysX 4 would be removed in development after the next release, but we're not committing to that at this stage, so it may be around longer.

Makes sense that we don't have a firm estimated time of removal at this point. But making sure we have an issue that documents an intent for removing PhysX 4 at the very least allows O3DE to indicate that it will not support PhysX 4 in perpetuity.

File: Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h
Comment (hultonha) [reply Code/Framework/AzFramework/AzFramework/Physics/Common/PhysicsSceneQueries.h:76]: Okay understood thanks for clarifying 👍

File: Gems/PhysX/Code/Source/Pipeline/MeshExporter.cpp
Comment (hultonha) [reply Gems/PhysX/Code/Source/Pipeline/MeshExporter.cpp:139]: Okay perfect, makes sense, thanks for the explanation

File: cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake
Comment (hultonha) [reply cmake/3rdParty/Platform/Mac/BuiltInPackages_mac.cmake:50]: Ah yeah you could be right, no problem 👍

**Review by hultonha — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp
Comment (SergeyAMZN) [reply Gems/PhysX/Code/Source/Common/PhysXSceneQueryHelpers.cpp:317]: Removed this comment since we can't actually delete/deprecate this function until the base class function is removed from PhysX SDK

**Comment by SergeyAMZN** [comment]
Deprecation issue for SIG-SIMULATION: https://github.com/o3de/o3de/issues/13897

**Comment by hultonha** [comment]
> Deprecation issue for SIG-SIMULATION: #13897

Awesome work @SergeyAMZN! Thanks very much 🙂

File: Gems/PhysX/Code/Tests/PhysXForceRegionTest.cpp
Comment (lumberyard-employee-dm) [inline Gems/PhysX/Code/Tests/PhysXForceRegionTest.cpp:320]: This is wrong, we should go from an EXPECT_GT to an EXPECT_TRUE when comparing to floats.

Before when this test failed it would output the `entityVelocity.GetZ()` value.

Now when it fails it just outputs that the expression "entityVelocity.GetZ() > -12.66f" has failed without being able to resolve the variable values.

```suggestion
        EXPECT_GT(entityVelocity.GetZ(), -12.66f); // Falling velocity should be slower than free fall velocity, which is -12.65 (+-0.01).
```

Comment (lawsonamzn) [reply Gems/PhysX/Code/Tests/PhysXForceRegionTest.cpp:320]: You can also add any amount of additional context you want since the EXPECT_xxxx macros return a stream that will be printed to the console, ie
```
EXPECT_xxxx(a, b) << "The acutal value of blah is " << blah; 
```

Comment (SergeyAMZN) [reply Gems/PhysX/Code/Tests/PhysXForceRegionTest.cpp:320]: nice spot! 
I actually made them to be EXPECT_GT/LT exactly for that reason, unit test was failing but no visibility what values were: https://github.com/o3de/o3de/pull/13802/commits/a84ea6aa584a829d5edfd633043d49fca02a701e

Must be switching between branches locally caused the bad merge. 

Thanks!

**Review by lumberyard-employee-dm — APPROVED** [review]

**Pull request merged**
